### PR TITLE
Permit realizing over smaller sizes than minimum by cropping output

### PR DESCRIPTION
**Pull request by abadams**
This is the 'view' solution from #3883 , though it could be easily adapted to be the copy solution too.

It avoids error messages about writing out of bounds on the output when calling realize with int args. It does an output bounds query, realizes over the upgraded domain, and then returns a crop back to the requested size. This crop may not be contiguous in memory.

### CODE REVIEW

**Comment by abadams** [comment]
This change revealed a bug where debug_to_file gets called on outputs even if they're in bounds query mode, so there's a fix for that here too.

**Comment by alexreinking** [comment]
Does this work with split domains? e.g. RDom over [5,10]x[5,10] but a realization over [0,7]x[0x7]? That realization is strictly speaking larger than the minimum, but not in the "right" position.

**Comment by abadams** [comment]
It will realize over the union of the box requested and the box written to, so you'd get a 10x10 output that gets cropped to the top-left 7x7.

**Comment by alexreinking** [comment]
Cool. That's what I understood but wanted to confirm.

**Comment by alexreinking** [comment]
What is causing the failures? Something I can help with?

**Comment by abadams** [comment]
When you make a bounds query call to ask a pipeline about output shapes (which we now do in this PR) it transitively makes bounds queries to any extern stages. My guess is that there's a test with an extern stage that is doing something funny that needs to be accounted for. I'll take a look on Monday.

**Comment by alexreinking** [comment]
Yeah, the extern stage in `skip_stages_external_array_functions` is incrementing a counter for how many bounds queries are made to it. The up-front bounds query adds a couple.

**Comment by abadams** [comment]
simd_op_check is compiled with NoBoundsQuery, so it's not going to work with this path. I'll add code to revert to the existing behavior if we're got NoBoundsQuery on. This will also give people a way to disable this new behavior if they don't want a bounds query for some reason.

**Comment by alexreinking** [comment]
We should also update the documentation in Func.h to make a note of the new behavior.

**Comment by abadams** [comment]
> We should also update the documentation in Func.h to make a note of the new behavior.

done. Take a look and see if it makes sense to you.

File: src/Func.h [hunk 1]
```diff
@@ -758,6 +758,16 @@ class Func {
...
+     * If the Func cannot be realized into a buffer of the given size
+     * due to scheduling constraints on scattering update definitions,
+     * it will be realized into a larger buffer of the minimum size
+     * possible, and a cropped view at the requested size will be
+     * returned. It is thus not safe to assume the returned buffers
+     * are contiguous in memory. This behavior can be disabled with
+     * the NoBoundsQuery target flag, in which case an error about
+     * writing out of bounds on the output buffer will trigger
+     * instead.
+     *
```
Comment (alexreinking) [inline src/Func.h]: Sounds good to me!

**Comment by steven-johnson** [comment]
Is this a WIP or is it ready to land (pending green)?

**Comment by abadams** [comment]
It's ready to land pending green.

**Comment by steven-johnson** [comment]
Failure is unrelated.

**Pull request merged**